Add development docs regarding testing (#4552)
kubernetes-sigs/kueue: site/content/en/docs/tasks/dev/testing.md

Reviewer: Also it would be great to mention `make verify`.

Reviewer: It would be great to mention how to change `E2E_KIND_VERSION`.

Reviewer 2: Actually, initially it was put under "Contribution guidelines", but I thought "Developer Tools" make sense. 

However, I'm ok either way. Contribution guide seems ok to me too.

Reviewer 2: @tenzen-y I'm ok to move the new page under contribution guidelines, are you ok with that?

Author: I moved it back to contribution guidelines

Reviewer 3: Thank you. I'm ok with that!

---

> title: "Running tests"

Reviewer 2: ```suggestion
title: "Running and debugging tests"
```

linkTitle: "Running tests"
weight: 25
description: >

>   Running tests in Kueue

Reviewer 2: ```suggestion
  Running and debugging tests
```
nit

---

## Running presubmission verification tests
```shell
make verify
```

## Running unit tests
To run all unit tests:
```shell
make test
```

To run unit tests for webhooks:
```shell
go test ./pkg/webhooks
```
To run tests that match `TestValidateClusterQueue` regular expression [ref](https://pkg.go.dev/cmd/go#hdr-Testing_flags):
```shell
go test ./pkg/webhooks -run TestValidateClusterQueue
```

### Running unit tests with race detection

```shell
# install go stress
go install golang.org/x/tools/cmd/stress@latest
# recompile with race instrumentation
go test ./pkg/scheduler/preemption/ -race -c
# it loops and reports failures
stress ./preemption.test -test.run TestPreemption

> ```

Reviewer: I think we missed this?

> ```

Author: @mbobrovskyi I pushed the updated version

## Running integration tests

For running subset of tests see [Running subset of tests](#running-subset-of-integration-or-e2e-tests).

### Increase logging verbosity
You can change log level in the framework (for example, change -3 to -5 to increase verbosity):
https://github.com/kubernetes-sigs/kueue/blob/f8015cb273f9115c34f9be32b35f7e1308c16459/test/integration/framework/framework.go#L72:
```go
var setupLogger = sync.OnceFunc(func() {
	ctrl.SetLogger(util.NewTestingLogger(ginkgo.GinkgoWriter, -3))
})
```

## Running e2e tests using custom build
```shell

> make kind-image-build test-e2e

Reviewer: Also we need to mention how to test other e2e tests

```suggestion
make kind-image-build 
make test-e2e 
make test-tas-e2e 
make test-e2e-customconfigs 
make test-multikueue-e2e
```

```

For running subset of tests see [Running subset of tests](#running-subset-of-integration-or-e2e-tests).

## Debug tests in VSCode
It is possible to debug unit and integration tests in VSCode.
You need to have the [Go extension](https://marketplace.visualstudio.com/items?itemName=golang.Go) installed.
Now you will have `run test | debug test` text buttons above lines like
```go
func TestValidateClusterQueue(t *testing.T) {
```
You can click on the `debug test` to debug a specific test.

> For integration tests, an additional step is needed.  Run `make test-integration | grep KUBEBUILDER_ASSETS` and add the path to the variable in settings.json:

Reviewer: ```suggestion
For integration tests, an additional step is needed.  Run `ENVTEST_K8S_VERSION=1.32 make envtest && ./bin/setup-envtest use $ENVTEST_K8S_VERSION -p path` and add the path to the variable in settings.json:
```

Reviewer 4: this way it's faster :)

```json
"go.testEnvVars": {

>     "KUBEBUILDER_ASSETS": "<path from output above>",

Reviewer: If we are testing kueuectl tests, we need to mention that the `KUEUE_BIN` variable must be set.

  },
```

## Running subset of integration or e2e tests
### Use Ginkgo --focus arg
```shell
GINKGO_ARGS="--focus=Scheduler" make test-integration
GINKGO_ARGS="--focus=Creating a Pod requesting TAS" make test-e2e
```
### Use ginkgo.FIt
If you want to focus on a specific tests, you can change

> `ginkgo.It` to `ginkgo.FIt` for these tests [ref][https://onsi.github.io/ginkgo/#focused-specs].

Reviewer: ```suggestion
`ginkgo.It` to `ginkgo.FIt` for these tests. 
For more details, see [here](https://onsi.github.io/ginkgo/#focused-specs).
```

Then the other tests will be skipped.
For example, you can change
```go
ginkgo.It("Should place pods based on the ranks-ordering", func() {
```
to
```go
ginkgo.FIt("Should place pods based on the ranks-ordering", func() {
```
and then run
```shell
# build and pull image
make test-tas-e2e
```
to test a particular TAS e2e test.

### Use INTEGRATION_TARGET
```shell
INTEGRATION_TARGET='test/integration/singlecluster/scheduler' make test-integration
```

## Flaky integration/e2e tests
You can use a script to repeatedly test:
```shell

> echo "going to test $NUM_TEST_ITERATIONS times"

Reviewer 4: I don't think the script is needed, as we can use ginkgo related decorators and flags e.g:
* MustPassRepeatedly
* FlakeAttempts
More here: https://onsi.github.io/ginkgo/#repeating-spec-runs-and-managing-flaky-specs

for i in $(seq 1 $NUM_TEST_ITERATIONS); do
  GINKGO_ARGS="--focus=Scheduler" make test-integration >> loop-test.log 2>&1
  echo iteration $i
  grep -q "FAIL!" loop-test.log
  if [ $? -eq 0 ]; then
    echo "Test run contains FAILED tests"
    exit 1
  fi
done
echo "looped the test $NUM_TEST_ITERATIONS times"
```

### Adding stress
You can run `stress` to increase CPU load during tests.
```shell
# install stress, for example with

> sudo apt install stress

Reviewer 3: Does this indicate https://packages.ubuntu.com/noble/stress (https://github.com/resurrecting-open-source-projects/stress)?

Author: @tenzen-y yes

Reviewer 3: Not all developers use Linux machines (as an alternative to MacOS). So, it would be nice to mention the tool link.
Could you add the tool link here?

# run stress alongside tests

> /usr/bin/stress --cpu 80

Reviewer: I think this command won't work.

Author: @mbobrovskyi  do you mean it's useless in context of stress-testing integration/e2e tests?

Reviewer: Oh, I see. It's another utility. Please ignore my comment.

```

And you can add --race to Ginkgo arguments for e2e tests (on integration tests it's already used):
```shell
GINKGO_ARGS="--race --focus=Scheduler" make test-e2e

> ```

Reviewer: I'm not sure how it is useful for e2e tests.

Reviewer 2: yeah, I haven't actually used it for E2e tests in the past. let's drop until proven useful.

Reviewer 4: Really useful is actually to attach to existing kind cluster, usually what I do is to comment the last line of hack/*e2e-tst.sh and later run tests with Ginkgo Tests plugin in VS code.
We might want to help ourselves in the future and maybe provide a flag to do it without modifying script every time ;)

Author: > Really useful is actually to attach to existing kind cluster, usually what I do is to comment the last line of hack/*e2e-tst.sh (run make test-*e2e) and later run tests with Ginkgo Tests plugin in VS code. We might want to help ourselves in the future and maybe provide a flag to do it without modifying script every time ;)

I'm not sure I understood.  Can you provide an example?  I've tried to run Ginkgo tests from VSCode, but then I got errors about `etcd` missing.

Reviewer 4: Let me know if that worked for you

Author: Thanks, it worked, at least from commandline, updated the docs.

go install golang.org/x/tools/cmd/stress@latest
# recompile with race instrumentation
go test ./pkg/scheduler/preemption/ -race -c
# it loops and reports failures
stress ./preemption.test -test.run TestPreemption
```

### Running integration tests

For running subset of tests see [Running subset of tests](#running-subset-of-integration-or-e2e-tests).

#### Increase logging verbosity
You can change log level in the framework (for example, change -3 to -5 to increase verbosity):
https://github.com/kubernetes-sigs/kueue/blob/f8015cb273f9115c34f9be32b35f7e1308c16459/test/integration/framework/framework.go#L72:
```go
var setupLogger = sync.OnceFunc(func() {
	ctrl.SetLogger(util.NewTestingLogger(ginkgo.GinkgoWriter, -3))
})
```

### Running e2e tests
```shell
# build an image
export IMAGE_REGISTRY=<image registry to which you have push access>
make image-local-push
# in the output you can image path that looks like the following:
# <image registry>/kueue:<tag>
# or, you can get it by combining $IMAGE_REGISTRY:`git describe --tags --dirty --always`
# pull that image:
docker pull <image registry>/kueue:<tag>
# run e2e test:
make test-e2e
# make test-tas-e2e
# make test-multikueue-e2e

> ```

Reviewer: Instead of this it's much easier to use `make kind-image-build test-e2e`

Reviewer 2: Or we may add a bit more details:

Step 1:
```sh
make kind-image-build
```
Step 2 against a new cluster:
```sh
IMAGE_TAG=us-central1-docker.pkg.dev/k8s-staging-images/kueue/kueue:<image version> CREATE_KIND_CLUSTER=true E2E_K8S_VERSIONS=1.32.0 make test-e2e
```
Or, step 2 against a pre-existing cluster:
```sh
kind load docker-image us-central1-docker.pkg.dev/k8s-staging-images/kueue/kueue:<image version> --name kind
IMAGE_TAG=us-central1-docker.pkg.dev/k8s-staging-images/kueue/kueue:<image version> CREATE_KIND_CLUSTER=true E2E_K8S_VERSIONS=1.32.0 make test-e2e
```
